pkg/microservice/warpdrive/core/service/taskplugin/deploy_helm.go

```diff
@@ -0,0 +1,655 @@
+	}
+	if replacedValuesYaml == "" {
+		err = errors.Errorf("failed to set new image uri into service's values.yaml %s/%s",
+			p.Task.Namespace, p.Task.ServiceName)
+		return
+	}
+
+	// merge override values and kvs into service's yaml
+	mergedValuesYaml, err = helmtool.MergeOverrideValues(serviceValuesYaml, renderInfo.DefaultValues, renderChart.GetOverrideYaml(), renderChart.OverrideValues)
+	if err != nil {
+		err = errors.WithMessagef(
+			err,
+			"failed to merge override values %s",
+			renderChart.OverrideValues,
+		)
+		return
+	}
+
+	// replace image into final merged values.yaml
+	replacedMergedValuesYaml, err = replaceImage(mergedValuesYaml, replaceValuesMap)
+	if err != nil {
+		err = errors.WithMessagef(
+			err,
+			"failed to replace image uri into helm values %s/%s",
+			p.Task.Namespace, p.Task.ServiceName)
+		return
+	}
+	if replacedMergedValuesYaml == "" {
+		err = errors.Errorf("failed to set image uri into mreged values.yaml in %s/%s",
+			p.Task.Namespace, p.Task.ServiceName)
+		return
+	}
+
+	p.Log.Infof("final replaced merged values: \n%s", replacedMergedValuesYaml)
+
+	helmClient, err = helmtool.NewClientFromNamespace(pipelineTask.ConfigPayload.DeployClusterID, p.Task.Namespace)
+	if err != nil {
+		err = errors.WithMessagef(
+			err,
+			"failed to create helm client %s/%s",
+			p.Task.Namespace, p.Task.ServiceName)
+		return
+	}
+
+	releaseName := util.GeneHelmReleaseName(p.Task.Namespace, p.Task.ServiceName)
+
+	ensureUpgrade := func() error {
+		hrs, errHistory := helmClient.ListReleaseHistory(releaseName, 10)
+		if errHistory != nil {
+			// list history should not block deploy operation, error will be logged instead of returned
+			p.Log.Errorf("failed to list release history, release: %s, err: %s", releaseName, errHistory)
+			return nil
+		}
+		if len(hrs) == 0 {
+			return nil
+		}
+		releaseutil.Reverse(hrs, releaseutil.SortByRevision)
+		rel := hrs[0]
+
+		if rel.Info.Status.IsPending() {
+			return fmt.Errorf("failed to upgrade release: %s with exceptional status: %s", releaseName, rel.Info.Status)
+		}
+		return nil
+	}
+
+	err = ensureUpgrade()
+	if err != nil {
+		return
+	}
+
+	timeOut := p.TaskTimeout()
+	chartSpec := helmclient.ChartSpec{
+		ReleaseName: releaseName,
+		ChartName:   chartPath,
+		Namespace:   p.Task.Namespace,
+		ReuseValues: true,
+		Version:     renderChart.ChartVersion,
+		ValuesYaml:  replacedMergedValuesYaml,
+		SkipCRDs:    false,
+		UpgradeCRDs: true,
+		Timeout:     time.Second * time.Duration(timeOut),
+		Wait:        true,
+		Replace:     true,
+		MaxHistory:  10,
+	}
+
+	done := make(chan bool)
+	go func(chan bool) {
+		if _, err = helmClient.InstallOrUpgradeChart(ctx, &chartSpec); err != nil {
+			err = errors.WithMessagef(
+				err,
+				"failed to upgrade helm chart %s/%s",
+				p.Task.Namespace, p.Task.ServiceName)
+			done <- false
+		} else {
+			done <- true
+		}
+	}(done)
+
+	select {
+	case <-done:
+		break
+	case <-time.After(chartSpec.Timeout + time.Minute):
```

Comment:
Why wait an extra minute?

Reply:
If the connection to the k8s cluster disconnected during deploying, the helm operation may stuck forever